File: docs/research/corpus-examples/file-handling/node/idiomatic-1.js
/*
 * Excerpt from Kelvi077/Ecommerce-Website-Template-FullStack, app.js:
 * multer storage/fileFilter config with filename sanitization and a
 * mimetype allowlist. See manifest.yaml for repo/commit/license
 * provenance. Trimmed to the upload-config lines; unrelated routes
 * omitted (`path` and `multer` are required earlier in the source file).
 */

// Set up file storage using multer
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, "public/images"); // Upload images to 'public/images' folder
  },
  filename: (req, file, cb) => {
    // SECURITY: Sanitize filenames to prevent directory traversal attacks
    const sanitizedFilename = path
      .basename(file.originalname)
      .replace(/[^a-zA-Z0-9_.-]/g, "_");
    const ext = path.extname(sanitizedFilename); // Get file extension
    cb(null, Date.now() + ext); // Use current timestamp as filename
  },
});

// SECURITY: Add file validation to multer
const fileFilter = (req, file, cb) => {
  // Accept only image files
  if (file.mimetype.startsWith("image/")) {
    cb(null, true);
  } else {
    cb(new Error("Only image files are allowed!"), false);
  }
};

// Initialize multer with storage configuration and file filter
const upload = multer({
  storage,
  fileFilter,
  limits: {
    fileSize: 5 * 1024 * 1024, // Limit file size to 5MB
  },
});

File: docs/research/corpus-examples/file-handling/python/idiomatic-1.py
"""
Excerpt from xuchengli/ai-answerer, app.py: `/upload` and `/uploads/<name>`
routes. Uses werkzeug's `secure_filename()` before `file.save()`, and
`send_from_directory()` (which itself confines the served path to
`UPLOAD_FOLDER`) for the matching download side. See manifest.yaml for
repo/commit/license provenance. Trimmed to the file-handling routes only;
unrelated TTS/LLM-answer routes and config (including a hardcoded demo
Redis URL/password in the source's `app.config` block) are omitted.
"""

import os
from flask import Flask, render_template, request, url_for, send_from_directory
from werkzeug.utils import secure_filename

app = Flask(__name__)
app.config['UPLOAD_FOLDER'] = 'uploads'


@app.post('/upload')
def upload():
    file = request.files['file']
    if file:
        filename = secure_filename(file.filename)
        file.save(os.path.join(app.config['UPLOAD_FOLDER'], filename))
    return url_for('download_file', name=filename)

@app.route('/uploads/<name>')
def download_file(name):
    return send_from_directory(app.config["UPLOAD_FOLDER"], name)
